README
    
    Copyright (C) 2011 Geodesic Limited, Mumbai, India

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
    INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
    PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE
    FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
    OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.

    Except as contained in this notice, the name of the Geodesic Limited shall not be used
    in advertising or otherwise to promote the sale, use or other dealings in this Software
    without prior written authorization from the Geodesic Limited.

-------------------------------------------------------------------------------------------

Aadhaar Auth-xml
----------------

* Source code for generating Aadhaar-AuthXML as per 1.5 specification of Aadhaar.
* Using source code, can generate Demographic and Biometric information 

Dependencies Libraries:
----------------------
	* openssl 0.9.8 or above
	* libxml	
	* xmlsec1 1.2.9 or above (if digital signature need to be generated)

API calls in aadhaar.c
----------------------
   + PID Xml generation
 	* pidxml_biometric
	* pidxml_demographic

   + Authxml generation
   	* Authxml_biometric
   	* Authxml_demographic

   + Digital Signature
	* do_digital_signature 
	
The above calls returns appropriate xml data.

XML Security:
------------
 * v1.2.9 used
 * Dependency : libxslt library


changes
Changes made after Initial Release:
----------------------------------
* Nov 22 2011:
	* Added one more API, authxml_biometric_with_fdc. This accepts fdc attribute value additionally.
	* Changes made in pidxml_biometric. Added <Meta> tag, and fdc attribute.

-------------------------------xxxxxxx--------------------------------------:


src/aadhaar.h
/*
    Copyright (C) 2011 Geodesic Limited, Mumbai, India

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
    INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
    PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE
    FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
    OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.

    Except as contained in this notice, the name of the Geodesic Limited shall not be used
    in advertising or otherwise to promote the sale, use or other dealings in this Software
    without prior written authorization from the Geodesic Limited.

*/

/*! 	\file aadhaar.h
*/

/*! \mainpage Aaddhaar Auth XML Generation API - Version 1.5
 *
 * \section intro_sec Introduction
 *
 * This documentation contains the details about how to use the Aadhaar API's to generate Auth XML version 1.5 in your Application.
 *
 *
 *
 * Dependencies:
 *
 *	The following libraries are required.
 *
 *	* Openssl 0.9.8 or above 
 *
 *	* LibXml 2.7.6 or above 
 *
 *
 * Additional Dependencies for Digital Signature Generation:
 * 
 * 	* libXslt
 *
 *  * libXmlSec1
 *
 * Note:
 *  
 *	You have to include the header file aadhaar.h in your application to use API's.
 *   
 *	While compiling your application you have to link the library -lcrypto -lxml2.
 *
 *  If Digital Signature used, then -lxslt -lxmlsec1 should be linked.
 *
 * 	example: gcc filename.c <linking_libraries>
 *
 * *
 */
#ifndef __AADHAAR_H__
#define __AADHAAR_H__

#ifdef __cplusplus
extern "C" {
#endif

/*! \def SHA256_LENGTH
 * \brief Sha 256 Length, used by Hmac function to generate Hash value.
 */
#define SHA256_LENGTH 32

/*! \def LOG_FILE
 * \brief log file location to store logs for debugging purpose.
 */
#define LOG_FILE "/tmp/log.txt"
/*! \def LICENCE_KEY_ONE
 * \brief Licence Key, Used by AuthXml lk attribute.
 */
#define LICENCE_KEY_ONE "MKg8njN6O+QRUmYF+TrbBUCqlrCnbN/Ns6hYbnnaOk99e5UGNhhE/xQ="

char duid[16];
char fdc[11]; // fdc attribute

/*!\fn unsigned char * pidxml_demographic(char *pname)
 * \brief This function is used to generate demographic PID XML data by sending Name as parameter. Returns XML data
 *
 * \param pname.
 *
 * \return xmldata.
 */
unsigned char * pidxml_demographic(char *pname);
/*!\fn unsigned char * pidxml_biometric(char *tmpldata)
 * \brief This function is used to generate biometric PID XML data by sending encoded fingerprint template as parameter. Returns XML data
 *
 * \param tmpldata.
 *
 * \return xmldata.
 */
unsigned char * pidxml_biometric(char *tmplData);
/*!\fn unsigned char * authxml_demographic_details(char *puid, char *pname)
 * \brief This function is used to generate Demographic Auth XML by sending parameter values of Aadhaar id and Name. Returns XML data
 *
 * \param puid
 * \param pname.
 *
 * \return xmldata.
 */
unsigned char * authxml_demographic_details(char *puid, char *pname);
/*!\fn unsigned char * authxml_biometric(char *puid, char *tmplData)
 * \brief This function is used to generate Biometric Auth XML by sending parameter values of Aadhaar id and Encoded fingerprint template data. Returns XML data
 *
 * \param puid
 * \param tmplData.
 *
 * \return xmldata.
 */
unsigned char * authxml_biometric(char *puid, char *tmplData);	
/*!\fn unsigned char * authxml_biometric_with_fdc(char *puid, char *pfdc, char *tmplData)
 * \brief This function is used to generate Biometric Auth XML by sending parameter values of Aadhaar id, fdc value and Encoded fingerprint template data. Returns XML data
 *
 * \param puid
 * \param pfdc
 * \param tmplData
 *
 * \return xmldata.
 */
unsigned char * authxml_biometric_with_fdc(char *puid, char *pfdc, char *tmplData);


/*!\fn int validate_uid(char *uId)
 * \brief This function is used to validate Aadhaar ID by sending parameter value of Aadhaar id. Return value 0 or -1.
 *
 * \param uId
 *
 * \return value. 0 for Success, -1 for failure.
 */
int validate_uid( char *uId );

#ifdef __cplusplus
}
#endif

#endif


src/uid_auth.h
/*
    Copyright (C) 2011 Geodesic Limited, Mumbai, India

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
    INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
    PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE
    FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
    OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.

    Except as contained in this notice, the name of the Geodesic Limited shall not be used
    in advertising or otherwise to promote the sale, use or other dealings in this Software
    without prior written authorization from the Geodesic Limited.

*/



#ifndef __UID_AUTH__
#define __UID_AUTH__
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/rsa.h>
#include <openssl/evp.h>
#include <openssl/objects.h>
#include <openssl/x509.h>
#include <openssl/err.h>
#include <openssl/pem.h>
#include <openssl/ssl.h>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/sha.h>
#include <openssl/bn.h>
#include <openssl/asn1.h>
#include <openssl/pkcs12.h>

#ifdef SERVER_PRODUCTION
#define UIDAI_PUBLIC_CERTIFICATE "/tmp/alive_auth/uid_prod.cer"
#define AUA_PRIVATE_CERTIFICATE "/tmp/public.p12"
#else
#define UIDAI_PUBLIC_CERTIFICATE "/tmp/uidai_auth_stage.cer"
#define AUA_PRIVATE_CERTIFICATE "/tmp/public.p12"
#endif

#define MAX_PUBKEY_SIZE   256
#define RSA_KEY_OFFSET    27
#define AES_KEY_LEN       32  // Changed according to spec 1.5 rev 1
#define RSA_KEY_LEN       32  // Changed according to spec 1.5 rev 1

#define STORE_TEMP_AUTH_XML "auth-xml.xml"
#define TEMP_STORE_CERTIFICATE "/tmp/cer" // Data of X509 Certicate Tag
#define TEMP_STORE_KEY_CERTIFICATE "/tmp/pri" // Private Key PEM File

#define KEYPASS "public"

/*unsigned char *uid_get_skey_data (unsigned char *key);
unsigned char *uid_get_aes_encrypted_data (unsigned char *in, int inlen, 
		 unsigned char *key);*/
#endif


src/aadhaar.c
/*
    Copyright (C) 2011 Geodesic Limited, Mumbai, India

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
    INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
    PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE
    FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
    OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
    DEALINGS IN THE SOFTWARE.

    Except as contained in this notice, the name of the Geodesic Limited shall not be used
    in advertising or otherwise to promote the sale, use or other dealings in this Software
    without prior written authorization from the Geodesic Limited.

*/

#include "aadhaar.h"
#include "uid_auth.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <libxml/parser.h>

extern char * get_expiry_date(char *expiry);
extern unsigned char *uid_get_skey_data (unsigned char *key);
extern unsigned char *uid_get_aes_encrypted_data (unsigned char *in, int inlen, 
		 unsigned char *key);
extern int uid_encode_b64(unsigned char *in, int inlen, unsigned char *out, 
			int *outlen);
			
/**********************************
	Validate Aadhaar ID
**********************************/
int validate_uid( char *uId )
{
	int dMultTable[10][10] = {{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
		{1, 2, 3, 4, 0, 6, 7, 8, 9, 5},
		{2, 3, 4, 0, 1, 7, 8, 9, 5, 6},
		{3, 4, 0, 1, 2, 8, 9, 5, 6, 7},
		{4, 0, 1, 2, 3, 9, 5, 6, 7, 8},
		{5, 9, 8, 7, 6, 0, 4, 3, 2, 1},
		{6, 5, 9, 8, 7, 1, 0, 4, 3, 2},
		{7, 6, 5, 9, 8, 2, 1, 0, 4, 3},
		{8, 7, 6, 5, 9, 3, 2, 1, 0, 4},
		{9, 8, 7, 6, 5, 4, 3, 2, 1, 0}};
	int permTable[8][10] = {{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
		{1, 5, 7, 6, 2, 8, 3, 0, 9, 4},
		{5, 8, 0, 3, 7, 9, 6, 1, 4, 2},
		{8, 9, 1, 6, 0, 4, 3, 5, 2, 7},
		{9, 4, 5, 3, 1, 2, 6, 8, 7, 0},
		{4, 2, 8, 6, 5, 7, 3, 9, 0, 1},
		{2, 7, 9, 3, 8, 0, 6, 4, 1, 5},
		{7, 0, 4, 6, 9, 1, 3, 2, 5, 8}};

	short int i=0, c=0;
	printf("validate_uid: %d:", c);
	for (i=0; i < 12; i++) {
		short int ni=0, newC=0;

		if (isdigit(uId[11-i]))
			ni = uId[11-i] - '0';
		else {
			printf("\n");
			return(-1);
		}
		newC = dMultTable[c][permTable[i%8][ni]];
		printf("%d:", newC);
		c = newC;
	}
	printf("\n");
	if (c == 0)
		return(0);
	return(-1);
}


char * parse_expiry_data(char *timestamp, char *expiry)
{
	
	char *strTimestamp=NULL;
	int n, k, mon, year, date;
	char m[][4] = {"Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};

	strTimestamp = strtok(timestamp," ");
	if (strTimestamp != NULL) {
		int j;

		for (j=0; j < 12; j++) {
			if (strcmp(m[j],strTimestamp) == 0) {
				mon = j+1;
				break;
			}
		}
	}
	strTimestamp = strtok(NULL," ");
	if (strTimestamp != NULL)
		date = atoi(strTimestamp);
	strTimestamp = strtok(NULL," ");
	strTimestamp = strtok(NULL," ");
	if (strTimestamp != NULL)
		year = atoi(strTimestamp);
        sprintf(expiry, "%d%02d%02d", year, mon, date);
#ifdef DEBUG	
	printf("parse_expiry_data: value is :%s:\n", expiry);
#endif
	return expiry;
}

unsigned char* pidxml_demographic(char *pname)
{
	xmlNodePtr root, demo, bios, n;
	xmlDocPtr doc;
	xmlChar *xmlbuff;
	int buffersize;
	struct tm * curr_tm;
        time_t curr_time;
	char buff[50];

        curr_time = time(NULL);
        curr_tm = localtime(&curr_time);
        sprintf(buff, "%04d-%02d-%02dT%02d:%02d:%02d", curr_tm->tm_year+1900, 
		curr_tm->tm_mon+1, curr_tm->tm_mday, 
		curr_tm->tm_hour, curr_tm->tm_min, curr_tm->tm_sec);

	doc = xmlNewDoc(NULL);
	root = xmlNewNode(NULL, "Pid");
	xmlSetProp(root, "ts", buff);
	xmlSetProp(root, "ver", "1.0");
	xmlSetProp(root, "xmlns", "http://www.uidai.gov.in/authentication/uid-auth-request-data/1.0");
	xmlDocSetRootElement(doc, root);

	demo = xmlNewNode(NULL, "Demo");
	n = xmlNewNode(NULL, "Pi");
	xmlSetProp(n, "ms", "E");
	xmlSetProp(n, "name", pname ? pname : "");
	xmlAddChild(demo, n);
	xmlAddChild(root, demo);

	xmlDocDumpFormatMemory(doc, &xmlbuff, &buffersize, 1);

	xmlFreeDoc(doc);
#if 1
	printf("\n############################################################\n%s\n", xmlbuff);
#endif
	return (unsigned char*)xmlbuff;
}

/***************************************************
	 Generate PID Xml - Biometric

	* tmplData - Encoded Biometric 
***************************************************/


unsigned char* pidxml_biometric(char *tmplData)
{
	xmlNodePtr root, demo, bios, n;
	xmlDocPtr doc;
	xmlChar *xmlbuff;
	int buffersize;
	struct tm * curr_tm;
        time_t curr_time;
	char buff[50];

        curr_time = time(NULL);
        curr_tm = localtime(&curr_time);
        sprintf(buff, "%04d-%02d-%02dT%02d:%02d:%02d", curr_tm->tm_year+1900, 
		curr_tm->tm_mon+1, curr_tm->tm_mday, 
		curr_tm->tm_hour, curr_tm->tm_min, curr_tm->tm_sec);

	doc = xmlNewDoc(NULL);
	root = xmlNewNode(NULL, "Pid");
	xmlSetProp(root, "ts", buff);
	xmlSetProp(root, "ver", "1.0");
	xmlSetProp(root, "xmlns", "http://www.uidai.gov.in/authentication/uid-auth-request-data/1.0");
	xmlDocSetRootElement(doc, root);
	
	if( fdc[0] != '\0')
	{
		n = xmlNewNode(NULL, "Meta");
		xmlSetProp(n, "fdc", fdc);
		xmlAddChild(root, n);
	}	

	bios = xmlNewNode(NULL, "Bios");
	n = xmlNewNode(NULL, "Bio");
	xmlSetProp(n, "type", "FMR");
	xmlNodeSetContent(n, tmplData);
	xmlAddChild(bios, n);
	xmlAddChild(root, bios);

	xmlDocDumpFormatMemory(doc, &xmlbuff, &buffersize, 1);

	xmlFreeDoc(doc);
	printf("\n############################################################\n%s\n", xmlbuff);
	return (unsigned char*)xmlbuff;

}

/***************************************************
	 Generate Biometric  - Auth Xml
***************************************************/

unsigned char * authxml_biometric(char *puid, char *tmplData)	
{
	xmlNodePtr root, n;
	xmlDocPtr doc;
	xmlChar *preDigSignedXmlBuff, *digSignedXmlBuff;
	int buffersize;
	char *pidb, *encryptedSessKey=NULL, *pload, *hmac;
	unsigned char sessKey[32], shaHash[65];
	unsigned char txnId[32], devId[16];

	sprintf(txnId, "%d", rand());
	strcpy(duid, puid);
	printf("\n\nUid value is :%s\n",duid);

	doc = xmlNewDoc("1.0");
	root = xmlNewNode(NULL, "Auth");
	xmlSetProp(root, "xmlns", "http://www.uidai.gov.in/authentication/uid-auth-request/1.0");
	xmlSetProp(root, "ver", "1.5");
	xmlSetProp(root, "tid", "public");
	xmlSetProp(root, "ac", "public");
	xmlSetProp(root, "sa", "public");
	xmlSetProp(root, "lk", LICENCE_KEY_ONE);
	xmlSetProp(root, "uid", puid ? puid : "");
	xmlSetProp(root, "txn", (const xmlChar *)txnId);
	xmlDocSetRootElement(doc, root);

	char bufExpiryStr[12];
	char *expiry=NULL;
	n = xmlNewNode(NULL, "Skey");
	xmlAddChild(root, n);
	{
		bzero(bufExpiryStr, 12);
		expiry = get_expiry_date(expiry);
		parse_expiry_data(expiry, bufExpiryStr);
		free(expiry);
		xmlSetProp(n, "ci", bufExpiryStr);
	}
	
	encryptedSessKey = uid_get_skey_data(sessKey);
	xmlNodeSetContent(n, encryptedSessKey);

	n = xmlNewNode(NULL, "Uses");
	xmlSetProp(n, "otp", "n"); //dpin as otp
	xmlSetProp(n, "pin", "n");
	xmlSetProp(n, "bio", "y");
	xmlSetProp(n, "pa", "n");
	xmlSetProp(n, "pfa", "n");
	xmlSetProp(n, "pi", "n");
	xmlSetProp(n, "bt", "FMR");
	xmlAddChild(root, n);

	pidb = pidxml_biometric(tmplData);
	n = xmlNewNode(NULL, "Data");
	xmlAddChild(root, n);
	pload = uid_get_aes_encrypted_data(pidb, strlen(pidb), sessKey);
	xmlNodeSetContent(n, pload);

	int res=hMacSha256(pidb, shaHash);

	hmac = uid_get_aes_encrypted_data(shaHash, SHA256_LENGTH, sessKey);
	n = xmlNewNode(NULL, "Hmac");
	xmlAddChild(root, n);
	xmlNodeSetContent(n, hmac);
	free(encryptedSessKey);
	free(hmac);

	xmlDocDumpFormatMemory(doc, &preDigSignedXmlBuff, &buffersize, 1);

	printf("\n\n AuthXML - 1:\n");
	printf("\n############################################################\n%s\n", preDigSignedXmlBuff);

#ifdef DEBUG	
	
	char str[7][256];

        sprintf(str[0], "echo size of base64 plain template buff is %d >> %s", strlen(tmplData), LOG_FILE);//Data in plain XML format
        system(str[0]);
        sprintf(str[1], "echo size of base64 template XML buff is %d >> %s", strlen(pidb), LOG_FILE);//Data in plain XML format
        system(str[1]);
        sprintf(str[2], "echo size of encrypted template buff is %d >> %s", strlen(pload), LOG_FILE);//Data in encrypted XML format
        system(str[2]);
        sprintf(str[3], "echo size of final encrypted xmlbuff is %d >> %s", strlen(preDigSignedXmlBuff), LOG_FILE); //final Data in Encrypted XML
        system(str[3]);
        sprintf(str[4], "echo -------------------------------------------------- >> %s", LOG_FILE);
        system(str[4]);
#endif	
	free(pidb);	
	free(pload);

#ifdef XML_SECURITY
	printf("\n############################################################\n");
	printf(" Digital Signature using XML Security\n\n");
	do_digital_signature(preDigSignedXmlBuff,&digSignedXmlBuff);
	if(preDigSignedXmlBuff)
		free(preDigSignedXmlBuff);
	xmlFreeDoc(doc);
	FILE *fp = fopen("/tmp/out.xml","w");
	fwrite(digSignedXmlBuff,1,strlen(digSignedXmlBuff),fp);
	fclose(fp);
	return ((unsigned char*)digSignedXmlBuff);
#else

	xmlFreeDoc(doc);
	return ((unsigned char*)preDigSignedXmlBuff);
#endif
}

/***************************************************
	 Generate Biometric  - Auth Xml with Fdc
***************************************************/

unsigned char * authxml_biometric_with_fdc(char *puid, char *pfdc, char *tmplData)	
{
	xmlNodePtr root, n;
	xmlDocPtr doc;
	xmlChar *preDigSignedXmlBuff, *digSignedXmlBuff;
	int buffersize;
	char *pidb, *encryptedSessKey=NULL, *pload, *hmac;
	unsigned char sessKey[32], shaHash[65];
	unsigned char txnId[32], devId[16];

	sprintf(txnId, "%d", rand());
	strcpy(duid, puid);
	printf("\n\nUid value is :%s\n",duid);

	// Fdc attribute value
	if (( pfdc != NULL ) && (strlen(pfdc) != 0))
		strcpy(fdc, pfdc);
	else
		fdc[0]='\0';

	doc = xmlNewDoc("1.0");
	root = xmlNewNode(NULL, "Auth");
	xmlSetProp(root, "xmlns", "http://www.uidai.gov.in/authentication/uid-auth-request/1.0");
	xmlSetProp(root, "ver", "1.5");
	xmlSetProp(root, "tid", "public");
	xmlSetProp(root, "ac", "public");
	xmlSetProp(root, "sa", "public");
	xmlSetProp(root, "lk", LICENCE_KEY_ONE);
	xmlSetProp(root, "uid", puid ? puid : "");
	xmlSetProp(root, "txn", (const xmlChar *)txnId);
	xmlDocSetRootElement(doc, root);

	char bufExpiryStr[12];
	char *expiry=NULL;
	n = xmlNewNode(NULL, "Skey");
	xmlAddChild(root, n);
	{
		bzero(bufExpiryStr, 12);
		expiry = get_expiry_date(expiry);
		parse_expiry_data(expiry, bufExpiryStr);
		free(expiry);
		xmlSetProp(n, "ci", bufExpiryStr);
	}
	
	encryptedSessKey = uid_get_skey_data(sessKey);
	xmlNodeSetContent(n, encryptedSessKey);

	n = xmlNewNode(NULL, "Uses");
	xmlSetProp(n, "otp", "n"); //dpin as otp
	xmlSetProp(n, "pin", "n");
	xmlSetProp(n, "bio", "y");
	xmlSetProp(n, "pa", "n");
	xmlSetProp(n, "pfa", "n");
	xmlSetProp(n, "pi", "n");
	xmlSetProp(n, "bt", "FMR");
	xmlAddChild(root, n);

	pidb = pidxml_biometric(tmplData);
	n = xmlNewNode(NULL, "Data");
	xmlAddChild(root, n);
	pload = uid_get_aes_encrypted_data(pidb, strlen(pidb), sessKey);
	xmlNodeSetContent(n, pload);

	int res=hMacSha256(pidb, shaHash);

	hmac = uid_get_aes_encrypted_data(shaHash, SHA256_LENGTH, sessKey);
	n = xmlNewNode(NULL, "Hmac");
	xmlAddChild(root, n);
	xmlNodeSetContent(n, hmac);
	free(encryptedSessKey);
	free(hmac);

	xmlDocDumpFormatMemory(doc, &preDigSignedXmlBuff, &buffersize, 1);

	printf("\n\n AuthXML - 1:\n");
	printf("\n############################################################\n%s\n", preDigSignedXmlBuff);

#ifdef DEBUG	
	
	char str[7][256];

        sprintf(str[0], "echo size of base64 plain template buff is %d >> %s", strlen(tmplData), LOG_FILE);//Data in plain XML format
        system(str[0]);
        sprintf(str[1], "echo size of base64 template XML buff is %d >> %s", strlen(pidb), LOG_FILE);//Data in plain XML format
        system(str[1]);
        sprintf(str[2], "echo size of encrypted template buff is %d >> %s", strlen(pload), LOG_FILE);//Data in encrypted XML format
        system(str[2]);
        sprintf(str[3], "echo size of final encrypted xmlbuff is %d >> %s", strlen(preDigSignedXmlBuff), LOG_FILE); //final Data in Encrypted XML
        system(str[3]);
        sprintf(str[4], "echo -------------------------------------------------- >> %s", LOG_FILE);
        system(str[4]);
#endif	
	free(pidb);	
	free(pload);

#ifdef XML_SECURITY
	printf("\n############################################################\n");
	printf(" Digital Signature using XML Security\n\n");
	do_digital_signature(preDigSignedXmlBuff,&digSignedXmlBuff);
	if(preDigSignedXmlBuff)
		free(preDigSignedXmlBuff);
	xmlFreeDoc(doc);
	FILE *fp = fopen("/tmp/out.xml","w");
	fwrite(digSignedXmlBuff,1,strlen(digSignedXmlBuff),fp);
	fclose(fp);
	return ((unsigned char*)digSignedXmlBuff);
#else

	xmlFreeDoc(doc);
	return ((unsigned char*)preDigSignedXmlBuff);
#endif
}


/***************************************************
	 Generate Demographic  - Auth Xml
***************************************************/

unsigned char* authxml_demographic_details(char *puid, char *pname)	
{
	xmlNodePtr root, n;
	xmlDocPtr doc;
	xmlChar *preDigSignedXmlBuff, *digSignedXmlBuff;
	int buffersize;
	char *pidb, *encryptedSessKey=NULL, *pload, *hmac;
	unsigned char sessKey[32];
	unsigned char txnId[32], devId[16];
	unsigned char shaHash[64];	

	sprintf(txnId, "%d", rand());
	strcpy(duid, puid);

	doc = xmlNewDoc("1.0");
	root = xmlNewNode(NULL, "Auth");
	xmlSetProp(root, "xmlns", "http://www.uidai.gov.in/authentication/uid-auth-request/1.0");
	xmlSetProp(root, "ver", "1.5");
	xmlSetProp(root, "tid", "public");
	xmlSetProp(root, "ac", "public");
	xmlSetProp(root, "sa", "public");
	xmlSetProp(root, "lk", LICENCE_KEY_ONE);
	xmlSetProp(root, "uid", puid ? puid : "");
	xmlSetProp(root, "txn", (const xmlChar *)txnId);
	xmlDocSetRootElement(doc, root);

	char bufExpiryStr[12];
	char *expiry=NULL;
	n = xmlNewNode(NULL, "Skey");
	{
		bzero(bufExpiryStr, 12);
		expiry = get_expiry_date(expiry);
		parse_expiry_data(expiry, bufExpiryStr);
		free(expiry);
		xmlSetProp(n, "ci", bufExpiryStr);
	}
	xmlAddChild(root, n);
	
	encryptedSessKey = uid_get_skey_data(sessKey);
	xmlNodeSetContent(n, encryptedSessKey);

	n = xmlNewNode(NULL, "Uses");
	xmlSetProp(n, "otp", "n");
	xmlSetProp(n, "pin", "n");
	xmlSetProp(n, "bio", "n");
	xmlSetProp(n, "pa", "n");
	xmlSetProp(n, "pfa", "n");
	xmlSetProp(n, "pi", "y");
	xmlAddChild(root, n);

	pidb = pidxml_demographic(pname);
	n = xmlNewNode(NULL, "Data");
	xmlAddChild(root, n);
	pload =  uid_get_aes_encrypted_data(pidb, strlen(pidb), sessKey);
	xmlNodeSetContent(n, pload);

	int res=hMacSha256(pidb, shaHash);

	hmac = uid_get_aes_encrypted_data(shaHash, SHA256_LENGTH, sessKey);
	n = xmlNewNode(NULL, "Hmac");
	xmlAddChild(root, n);
	xmlNodeSetContent(n, hmac);
	free(encryptedSessKey);
	free(hmac);

	xmlDocDumpFormatMemory(doc, &preDigSignedXmlBuff, &buffersize, 1);
	free(pidb);
	free(pload);

#ifdef DEBUG
	printf("\n\n AuthXML - 1:\n");
	printf("\n############################################################\n%s\n", preDigSignedXmlBuff);
#endif

#ifdef XML_SECURITY
	printf("\nDigital Signature using XML Security\n\n");
	do_digital_signature(preDigSignedXmlBuff,&digSignedXmlBuff);
	free(preDigSignedXmlBuff);
	xmlFreeDoc(doc);
	return((unsigned char*)digSignedXmlBuff);
#endif	

	xmlFreeDoc(doc);
	return((unsigned char*)preDigSignedXmlBuff);
}


int uid_auth_demographic_details(char *puid, char *name)
{
	unsigned char *authb=NULL;
	int res;
	
	authb = authxml_demographic_details(puid, name);
	puts(authb);
	free(authb);
	
	return res;
}

int main()
{
	authxml_demographic_details("123456789012","Shiva");
	return 0;
}
